relay-cardinality/src/redis.rs

```diff
@@ -50,171 +49,256 @@ impl RedisSetLimiter {
-        cache_hits.report(
-            CardinalityLimiterCounters::RedisCacheHit,
-            CardinalityLimiterCounters::RedisCacheMiss,
-        );
+        drop(cache); // Give up the cache lock!
 
         let mut client = self.redis.client()?;
         let mut connection = client.connection()?;
 
-        for ((organization_id, item_scope), entries) in entries_by_scope {
-            if entries.is_empty() {
+        for mut state in states {
+            if state.entries.is_empty() {
                 continue;
             }
 
-            let results = metric!(
-                timer(CardinalityLimiterTimers::Redis),
-                scope = &item_scope,
-                {
-                    self.check_limits(
-                        &mut connection,
-                        organization_id,
-                        item_scope.clone(),
-                        entries,
-                        timestamp,
-                        limit,
-                    )
-                }
-            )?;
+            let results = metric!(timer(CardinalityLimiterTimers::Redis), id = state.id, {
+                self.check_limits(&mut connection, &mut state, timestamp)
+            })?;
 
             // This always acquires a write lock, but we only hit this
             // if we previously didn't satisfy the request from the cache,
             // -> there is a very high chance we actually need the lock.
-            let mut cache = self
-                .cache
-                .update(organization_id, &item_scope, cache_window);
-
+            let mut cache = self.cache.update(state.scope, timestamp); // Acquire a write lock.
             for (entry, status) in results {
                 if status.is_rejected() {
-                    acceptions_rejections.miss(&item_scope);
                     rejected.push(entry.rejection());
+                    state.rejected();
                 } else {
-                    acceptions_rejections.hit(&item_scope);
                     cache.accept(entry.hash);
+                    state.accepted();
                 }
             }
+            drop(cache); // Give up the cache lock!
         }
 
         if !rejected.is_empty() {
             relay_log::debug!(
-                organization_id = organization_id,
+                scoping = ?scoping,
                 "rejected {} metrics due to cardinality limit",
                 rejected.len(),
             );
         }
 
-        acceptions_rejections.report(
-            CardinalityLimiterCounters::Accepted,
-            CardinalityLimiterCounters::Rejected,
-        );
-
         Ok(Box::new(rejected.into_iter()))
     }
 }
 
+/// A quota scoping extracted from a [`CardinalityLimit`] and a [`Scoping`].
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub(crate) struct QuotaScoping {
+    pub window: SlidingWindow,
+    pub namespace: Option<MetricNamespace>,
+    pub organization_id: Option<OrganizationId>,
+    pub project_id: Option<ProjectId>,
```

Comment:
nit: Are all combinations valid here? I.e. `(None, None, None)`, `(None, None, Some)`, etc.

Reply:
Yes they are, a `(None, None, None)` would be a total global cardinality limit, `(Some, Some, Some)` would be a project scoped limit for a certain namespace.

They only somewhat invalid config would be `(Some/None, None, Some)`, it works, just does not make sense, since a project will always have an org (hopefully).